File: 03-SOLID-Principles/src/go/isp.go
package main

import "fmt"

// ISP — Interface Segregation Principle

type Printable interface {
	Print(doc string)
}

type Scannable interface {
	Scan(doc string)
}

type Faxable interface {
	Fax(doc string)
}

type AllInOnePrinter struct{}
func (AllInOnePrinter) Print(doc string) { fmt.Println("Printing:", doc) }
func (AllInOnePrinter) Scan(doc string)  { fmt.Println("Scanning:", doc) }
func (AllInOnePrinter) Fax(doc string)   { fmt.Println("Faxing:", doc) }

type SimplePrinterISP struct{}
func (SimplePrinterISP) Print(doc string) { fmt.Println("Printing:", doc) }
// No Scan, no Fax — only satisfies Printable

File: 03-SOLID-Principles/src/rust/isp.rs
/// ISP — Interface Segregation Principle
/// Small focused traits instead of one fat trait

pub trait Printable {
    fn print(&self, doc: &str);
}

pub trait Scannable {
    fn scan(&self, doc: &str);
}

pub trait Faxable {
    fn fax(&self, doc: &str);
}

pub struct AllInOnePrinter;
impl Printable for AllInOnePrinter {
    fn print(&self, doc: &str) { println!("Printing: {}", doc); }
}
impl Scannable for AllInOnePrinter {
    fn scan(&self, doc: &str) { println!("Scanning: {}", doc); }
}
impl Faxable for AllInOnePrinter {
    fn fax(&self, doc: &str) { println!("Faxing: {}", doc); }
}

pub struct SimplePrinter;
impl Printable for SimplePrinter {
    fn print(&self, doc: &str) { println!("Printing: {}", doc); }
}
// SimplePrinter doesn't implement Scannable or Faxable — compiler enforces ISP
